rllib/models/specs/specs_base.py

```diff
@@ -0,0 +1,209 @@
+        >>> spec.validate(torch.rand(32, 128, dtype=torch.float64))  # raises ValueError
+
+    Public Methods:
+        validate: Checks if the shape and dtype of the tensor matches the
+            specification.
+        sample: Samples a tensor with the specified value that is an
+            example of a tensor that matches the specification.
+
+    Abstract Methods:
+        get_shape: Returns the shape of the tensor depending on the backend.
+        get_dtype: Returns the dtype of the tensor depending on the backend.
+        _sample: Samples a tensor with the specified value that
+            has values of fill_value, shape of shape, and dtype of self.dtype.
+    """
+
+    def __init__(
+        self, shape: str, *, dtype: Optional[Any] = None, **shape_vals: Dict[str, int]
+    ) -> None:
+        self._expected_shape = self._parse_expected_shape(shape, shape_vals)
+        self._dtype = dtype
+
+    @property
+    def shape(self) -> Tuple[int]:
+        """Returns a `tuple` specifying the tensor shape."""
+        return self._expected_shape
+
+    @property
+    def dtype(self) -> Any:
+        """Returns a dtype specifying the tensor dtype."""
+        return self._dtype
+
+    @DeveloperAPI
+    def validate(self, tensor: TensorType) -> None:
+        """Checks if the shape and dtype of the tensor matches the specification.
+
+        Args:
+            tensor: The tensor to be validated.
+
+        Raises:
+            ValueError: If the shape or dtype of the tensor does not match the
+        """
+        shape = self.get_shape(tensor)
+        if len(shape) != len(self._expected_shape):
+            raise ValueError(_INVALID_SHAPE.format(self._expected_shape, shape))
+
+        for expected_d, actual_d in zip(self._expected_shape, shape):
+            if isinstance(expected_d, int) and expected_d != actual_d:
+                raise ValueError(_INVALID_SHAPE.format(self._expected_shape, shape))
+
+        dtype = self.get_dtype(tensor)
+        if self.dtype and dtype != self.dtype:
+            raise ValueError(_INVALID_DTYPE.format(self.dtype, tensor.dtype))
+
+    @abc.abstractmethod
+    def get_shape(self, tensor: TensorType) -> Tuple[int]:
+        """Returns the shape of a tensor.
+
+        Args:
+            tensor: The tensor whose shape is to be returned.
+
+        Returns:
+            A `tuple` specifying the shape of the tensor.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def get_dtype(self, tensor: TensorType) -> Any:
+        """Returns the data type of a tensor.
+
+        Args:
+            tensor: The tensor whose data type is to be returned.
+
+        Returns:
+            The data type of the tensor.
+        """
+        raise NotImplementedError
+
+    @DeveloperAPI
+    def sample(self, fill_value: Union[float, int] = 0) -> TensorType:
+        """Samples a tensor with the specified value that is matches the specs.
+
+        Args:
```

Comment:
should this be called sample? That implies randomness, no? Maybe we call it create_from_value, or fill_new?

Reply:
It should be sample in a sense that you have variable sized dimensions that could take any value. One implementation of this function which we do here is to always sample 1 (a deterministic value) for TBD dimension sizes like Batch or time in RNN or channel in CNNs.

Reply:
I get that `sample` is in line with the gym interface, but maybe `full` would be more intuitive for pytorch/numpy users. We might want to implement domains (e.g. `Box(low=0, high=1.0)`) where we would actually want a `sample` method to return a uniform-random tensor.